api/v1beta1/conversion.go

```diff
@@ -16,14 +16,187 @@ limitations under the License.
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
 
-// Hub marks CloudStackClusterList as a conversion hub.
-func (*CloudStackClusterList) Hub() {}
+const DefaultEndpointCredential = "global"
 
-// Hub marks CloudStackMachine as a conversion hub.
-func (*CloudStackMachine) Hub() {}
+//nolint:golint,revive,stylecheck
+func Convert_v1beta1_CloudStackCluster_To_v1beta2_CloudStackCluster(in *CloudStackCluster, out *v1beta2.CloudStackCluster, s conv.Scope) error {
+	out.ObjectMeta = in.ObjectMeta
+	failureDomains, err := getFailureDomains(in)
+	if err != nil {
+		return err
+	}
+	out.Spec = v1beta2.CloudStackClusterSpec{
+		ControlPlaneEndpoint: in.Spec.ControlPlaneEndpoint,
+		FailureDomains:       failureDomains,
+		Account:              in.Spec.Account,
+		Domain:               in.Spec.Domain,
+	}
 
-// Hub marks CloudStackMachineList as a conversion hub.
-func (*CloudStackMachineList) Hub() {}
+	out.Status = v1beta2.CloudStackClusterStatus{
+		FailureDomains: in.Status.FailureDomains,
+		Ready:          in.Status.Ready,
+	}
+	return nil
+}
+
+//nolint:golint,revive,stylecheck
+func Convert_v1beta2_CloudStackCluster_To_v1beta1_CloudStackCluster(in *v1beta2.CloudStackCluster, out *CloudStackCluster, scope conv.Scope) error {
+	out.ObjectMeta = in.ObjectMeta
+	out.Spec = CloudStackClusterSpec{
+		Zones:                getZones(in),
+		ControlPlaneEndpoint: in.Spec.ControlPlaneEndpoint,
+		Account:              in.Spec.Account,
+		Domain:               in.Spec.Domain,
+	}
+
+	out.Status = CloudStackClusterStatus{
+		FailureDomains: in.Status.FailureDomains,
+		Ready:          in.Status.Ready,
+	}
+	return nil
+}
+
+// getZones maps failure domains to zones
+func getZones(csCluster *v1beta2.CloudStackCluster) []Zone {
+	var zones []Zone
+	for _, failureDomain := range csCluster.Spec.FailureDomains {
+		zone := failureDomain.Zone
+		zones = append(zones, Zone{
+			Name: zone.Name,
+			ID:   zone.ID,
+			Network: Network{
+				Name: zone.Network.Name,
+				ID:   zone.Network.ID,
+				Type: zone.Network.Type,
+			},
+		})
+	}
+	return zones
+}
+
+// getFailureDomains maps v1beta1 zones to v1beta2 failure domains.
+func getFailureDomains(csCluster *CloudStackCluster) ([]v1beta2.CloudStackFailureDomainSpec, error) {
+	var failureDomains []v1beta2.CloudStackFailureDomainSpec
+	namespace := csCluster.Namespace
+	for _, zone := range csCluster.Spec.Zones {
+		name, err := GetDefaultFailureDomainName(namespace, csCluster.Name, zone.ID, zone.Name)
+		if err != nil {
+			return nil, err
+		}
+		failureDomains = append(failureDomains, v1beta2.CloudStackFailureDomainSpec{
+			Name: name,
+			Zone: v1beta2.CloudStackZoneSpec{
+				ID:   zone.ID,
+				Name: zone.Name,
+				Network: v1beta2.Network{
+					ID:   zone.Network.ID,
+					Name: zone.Network.Name,
+					Type: zone.Network.Type,
+				},
+			},
+			Domain:  csCluster.Spec.Domain,
+			Account: csCluster.Spec.Account,
+			ACSEndpoint: corev1.SecretReference{
+				Namespace: csCluster.ObjectMeta.Namespace,
```

Comment:
Are csCluster.Namespace and csCluster.ObjectMeta.Namespace different?
If not, namespace can be used instead of csCluster.ObjectMeta.Namespace

Reply:
They are the same fields. csCluster.Namespace and csCluster.ObjectMeta.Namespace are two two ways to refer the same field since ObjectMeta is a anonynouse field.